source/game/kart/KartDynamics.cc

```diff
@@ -25,17 +49,75 @@ void KartDynamics::calc(f32 dt, f32 maxSpeed, bool /*air*/) {
     m_acceleration0 = m_totalForce;
     m_extVel += m_acceleration0 * dt;
     m_extVel *= 0.998f;
+    m_angVel0 *= 0.98f;
+
+    EGG::Vector3f playerBack = m_mainRot.rotateVector(EGG::Vector3f::ez);
+    EGG::Vector3f playerBackHoriz = playerBack;
+    playerBackHoriz.y = 0.0f;
+
+    if (FLT_EPSILON < playerBackHoriz.dot()) {
+        playerBackHoriz.normalise();
+        m_extVel = m_extVel.rej(playerBackHoriz);
+    }
+
     m_velocity = m_extVel * dt + m_intVel + m_movingObjVel + m_movingRoadVel;
     m_speedNorm = std::min(m_velocity.normalise(), maxSpeed);
     m_velocity *= m_speedNorm;
     m_pos += m_velocity;
 
+    EGG::Vector3f t1 = m_invInertiaTensor.multVector(m_totalTorque) * dt;
+    m_angVel0 += (t1 + m_invInertiaTensor.multVector(t1 + m_totalTorque) * dt) * 0.5f;
+
+    m_angVel0.x = std::min(0.4f, std::max(-0.4f, m_angVel0.x));
+    m_angVel0.y = std::min(0.4f, std::max(-0.4f, m_angVel0.y));
+    m_angVel0.z = std::min(0.8f, std::max(-0.8f, m_angVel0.z));
+    m_angVel0.y *= m_angVel0YFactor;
+
+    EGG::Vector3f angVelSum = m_angVel2 + m_angVel1 + m_angVel0Factor * m_angVel0;
+
+    if (FLT_EPSILON < angVelSum.dot()) {
+        m_mainRot += m_mainRot * angVelSum * dt * 0.5;
+
+        if (EGG::Mathf::abs(m_mainRot.dot()) < FLT_EPSILON) {
+            m_mainRot = EGG::Quatf::ident;
+        } else {
+            m_mainRot.normalise();
+        }
+    }
+
+    // TODO: Can we prove we don't need this given above???
+    if (EGG::Mathf::abs(m_mainRot.dot()) < FLT_EPSILON) {
+        m_mainRot = EGG::Quatf::ident;
+    } else {
+        m_mainRot.normalise();
+    }
```

Comment:
Yeah so this seemed like an unnecessary snippet. If the dot product being less than epsilon was a concern, it would have already been addressed and normalised in the previous if-statement, right?

Reply:
To match the base game, I'll keep this in.